### PR TITLE
Minor code cleanup

### PR DESCRIPTION
**Pull request by hahschaa**
### Prerequisites

- [x] I have verified that there are no overlapping [pull-requests](https://github.com/NetTopologySuite/NetTopologySuite/pulls) open
- [x] I have written a descriptive pull-request title
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository.

- [x] I have provided test coverage for my change (where applicable)

### Description

I checked out the repo to see if I could make the changes required for [issue 790](https://github.com/NetTopologySuite/NetTopologySuite/issues/790), but saw some minor code cleanup that I could apply.

### CODE REVIEW

**Comment by hahschaa** [comment]
By the way, I was kind of confused by the code in e.g. [BoundaryChainNoder.cs](https://github.com/NetTopologySuite/NetTopologySuite/pull/793/commits/06b665f97cc42c61995c77f408df3f7525ec04e1#diff-5a6d4057b4ebcf5dd578f9673e9cee41a25e5529a9dd4c087dc28dcaad12f13e), the method is called `AddSegments`, but if a segment already exists it is removed, and not re-added afterwards. Is that correct?

**Comment by hahschaa** [comment]
Sorry for the changes after opening the PR. I forgot that this project is also build for .net standard 2.0, so I had to revert some changes that are available in that target framework. But then I some more thing to clean.

**Comment by airbreather** [comment]
> By the way, I was kind of confused by the code in e.g. [BoundaryChainNoder.cs](https://github.com/NetTopologySuite/NetTopologySuite/pull/793/commits/06b665f97cc42c61995c77f408df3f7525ec04e1#diff-5a6d4057b4ebcf5dd578f9673e9cee41a25e5529a9dd4c087dc28dcaad12f13e), the method is called `AddSegments`, but if a segment already exists it is removed, and not re-added afterwards. Is that correct?

It feels like a correct **performance optimization** to me, and your change makes it **even better**.

Per the description in the summary, this is used to find "Boundary segments [...] which are not duplicated in the input polygonal coverage". The "input polygonal coverage" constraint means that any given segment will appear either once or twice in the entire coverage, never more.

It's also correct to use the return value of `HashSet<T>` methods to avoid a redundant `Contains` call (edit: I say "redundant" because both `Remove` and `Add` imply all the same work done by `Contains` — this is a generally applicable optimization that some tools will perform automatically). I don't expect it to be a HUGE win in practice, but it should never make anything worse, so I'm happy with it.

File: test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs [hunk 1]
```diff
@@ -56,11 +56,7 @@ public void RunCompare()
             var seg3 = CreateRandomDepthSegment();
 
             // do test work here
-            bool fails = false;
-            if (!IsSymmetric(seg1, seg2))
-                fails = true;
-            if (!IsTransitive(seg1, seg2, seg3))
-                fails = true;
+            bool fails = !IsSymmetric(seg1, seg2) || !IsTransitive(seg1, seg2, seg3);
```
Comment (airbreather) [inline test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs]: This is a behavior change. `IsTransitive` has side effects that will no longer run if `IsSymmetric` returns `false`, whereas they currently otherwise would run.

A fancy way to resolve this is to change `||` to `|`. A less fancy way is to leave it alone (edit to add:), i.e., revert this change.

File: src/NetTopologySuite/Utilities/HexConverter.cs [hunk 2]
```diff
@@ -22,78 +23,78 @@ private HexConverter() { }
...
-                  // generazione del risultato final
-                  // se il risultato da generare è in base 10 non c'è
-                  // bisogno di calcoli
-                  if (baseOut == 10)
-                           result = valore.ToString();
+                // generazione del risultato final
+                // se il risultato da generare è in base 10 non c'è
+                // bisogno di calcoli
+                if (baseOut == 10)
+                    result = valore.ToString(CultureInfo.InvariantCulture);
```
Comment (airbreather) [inline src/NetTopologySuite/Utilities/HexConverter.cs:79]: :thinking: I feel like this probably makes it worse.

This method also contains a `.ToUpper()` and a `double.Parse` which are culture-sensitive, and its two callers also both use unqualified `.ToString()`. Each of these uses the current culture implicitly.

I'm not sure it makes sense to change **only** this one. I would personally **at least** leave such a change for a more focused PR.

*The similar changes in 5234798 do not worry me today because they do not touch the `NetTopologySuite` package.*

File: src/NetTopologySuite/Algorithm/MinimumBoundingCircle.cs [hunk 3]
```diff
@@ -267,7 +267,7 @@ private void ComputeCirclePoints()
             // handle degenerate or trivial cases
             if (_input.IsEmpty)
             {
-                _extremalPts = new Coordinate[0];
+                _extremalPts = Array.Empty<Coordinate>();
```
Comment (airbreather) [inline src/NetTopologySuite/Algorithm/MinimumBoundingCircle.cs:270]: :page_facing_up: FYI this one in particular is technically a behavior change. I seriously doubt that this will ever matter, and I support the change, but I want to point out these subtleties when I see them:

Because `GetExtremalPoints()` returns this array directly, there will now be cases where `a.GetExtremalPoints() == b.GetExtremalPoints()` when they didn't before, even if `a` and `b` are different.

File: src/NetTopologySuite/Densify/Densifier.cs [hunk 4]
```diff
@@ -159,7 +159,7 @@ protected override CoordinateSequence TransformCoordinates(
                 // prevent creation of invalid LineStrings
                 if (parent is LineString && newPts.Length == 1)
                 {
-                    newPts = new Coordinate[0];
+                    newPts = Array.Empty<Coordinate>();
```
Comment (airbreather) [inline src/NetTopologySuite/Densify/Densifier.cs:162]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

File: src/NetTopologySuite/Geometries/Implementation/CoordinateArraySequence.cs [hunk 5]
```diff
@@ -64,7 +64,7 @@ public CoordinateArraySequence(Coordinate[] coordinates, int dimension, int meas
         {
             if (coordinates == null)
             {
-                Coordinates = new Coordinate[0];
+                Coordinates = Array.Empty<Coordinate>();
```
Comment (airbreather) [inline src/NetTopologySuite/Geometries/Implementation/CoordinateArraySequence.cs:67]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

File: src/NetTopologySuite/Geometries/Implementation/CoordinateArraySequence.cs [hunk 6]
```diff
@@ -119,7 +119,7 @@ public CoordinateArraySequence(CoordinateSequence coordSeq)
         {
             if (coordSeq == null)
             {
-                Coordinates = new Coordinate[0];
+                Coordinates = Array.Empty<Coordinate>();
```
Comment (airbreather) [inline src/NetTopologySuite/Geometries/Implementation/CoordinateArraySequence.cs:122]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

File: src/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequence.cs [hunk 7]
```diff
@@ -375,7 +375,7 @@ public PackedFloatCoordinateSequence(float[] coords, int dimension, int measures
             : base(coords?.Length / dimension ?? 0, dimension, measures)
         {
             if (coords == null)
-                coords = new float[0];
+                coords = Array.Empty<float>();
```
Comment (airbreather) [inline src/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequence.cs:378]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

File: src/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequence.cs [hunk 8]
```diff
@@ -395,7 +395,7 @@ public PackedFloatCoordinateSequence(double[] coords, int dimension, int measure
         {
             if (coords == null)
             {
-                _coords = new float[0];
+                _coords = Array.Empty<float>();
```
Comment (airbreather) [inline src/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequence.cs:398]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

File: src/NetTopologySuite/IO/WKTReader.cs [hunk 9]
```diff
@@ -865,7 +865,7 @@ private MultiPoint ReadMultiPointText(TokenStream tokens, GeometryFactory factor
             string nextToken = GetNextEmptyOrOpener(tokens);
             if (nextToken.Equals(WKTConstants.EMPTY))
             {
-                return factory.CreateMultiPoint(new Point[0]);
+                return factory.CreateMultiPoint(Array.Empty<Point>());
```
Comment (airbreather) [inline src/NetTopologySuite/IO/WKTReader.cs:868]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

File: src/NetTopologySuite/Simplify/VWSimplifier.cs [hunk 10]
```diff
@@ -115,7 +115,7 @@ protected override CoordinateSequence TransformCoordinates(CoordinateSequence co
                 var inputPts = coords.ToCoordinateArray();
                 Coordinate[] newPts;
                 if (inputPts.Length == 0)
-                    newPts = new Coordinate[0];
+                    newPts = Array.Empty<Coordinate>();
```
Comment (airbreather) [inline src/NetTopologySuite/Simplify/VWSimplifier.cs:118]: :page_facing_up: Same basic note about behavior change as in `MinimumBoundingCircle.cs`

**Review by airbreather — CHANGES_REQUESTED** [review]

File: test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs (hunk 1 above)
Comment (hahschaa) [reply test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs]: Which side effects do you mean? I only see some compares and a Console.WriteLine. As far as I can see the CompareTo only does integer comparisons without side effects. Technically the console output is a side effect, but is that really important if you already know that IsSymmetric has failed. Furthermore this is in a test, so it should have no influence on any actual operations.

File: src/NetTopologySuite/Algorithm/MinimumBoundingCircle.cs (hunk 3 above)
Comment (hahschaa) [reply src/NetTopologySuite/Algorithm/MinimumBoundingCircle.cs:270]: This is certainly true, but when comparing references types like arrays, it is important to always consider if you want reference equality versus value equality. My feeling is that when you are comparing a collection of Coordinates, you always want to do some value equality, and only want to check for reference equality in some edge cases.

File: test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs (hunk 1 above)
Comment (airbreather) [reply test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs]: > Technically the console output is a side effect, but is that really important if you already know that IsSymmetric has failed.

:shrug:

I'm not the one who wrote the `Console.WriteLine` call. I only assume that somebody added it for a reason. I do not know what the reason was, and I am not in a position to spend my time guessing.

[Without knowing that reason, I necessarily cannot claim that the behavior change is appropriate.](https://fs.blog/chestertons-fence)

> Furthermore this is in a test, so it should have no influence on any actual operations.

I have very little doubt that you are correct that this would never matter. I'm also not the one who went out of my way to propose a change to it...

File: src/NetTopologySuite/Utilities/HexConverter.cs (hunk 2 above)
Comment (hahschaa) [reply src/NetTopologySuite/Utilities/HexConverter.cs:79]: Whoops, I indeed missed the `ToUpper` and `Parse`. But how does it make things worse?

If I also change those to `.ToUpperInvariant()` and `double.Parse(..., CultureInfo.InvariantCulture);` then the whole thing is not dependent on the locale anymore. I think that is a good thing to have in a non-UI library?

Comment (airbreather) [reply src/NetTopologySuite/Utilities/HexConverter.cs:79]: To put my comment into context: I'm reviewing a PR entitled "Minor code cleanup", which touches 310 files in ways that are mostly harmless, mostly welcome, and mostly beneficial.

In such a PR, I expected to see **NO** behavior changes; if there **are** any behavior changes, then I was expecting them all to be of the `new T[0]` --> `Array.Empty<T>()` sort, which — while I believe it would not technically be correct for the .NET JIT to apply the same optimization — I'm not going to complain about, other than to call attention to them.

Bringing it back to my specific comment here, now.

---

> Whoops, I indeed missed the `ToUpper` and `Parse`.

And `ToString()` in both callers

> But how does it make things worse?

I don't know of any **specific, real-world** situations where the value of `CurrentCulture` would cause this to behave incorrectly (where it previously behaved correctly). That is why I said " :thinking: I feel like this probably makes it worse" instead of " :warning: This makes it worse".

Mixing `CultureInfo` throughout a single method (and its callers) needs to be justified. Continuing to use a consistent `CultureInfo` throughout a method that has stood as-is for roughly 20 years (or more?), on the other hand, does not IMO need to be justified (though if you **do** think that this should change, then it probably would not be hard to convince me to agree with you, and I would appreciate a **separate** issue and PR with more details about this situation).

> If I also change those to `.ToUpperInvariant()` and `double.Parse(..., CultureInfo.InvariantCulture);` then the whole thing is not dependent on the locale anymore. I think that is a good thing to have in a non-UI library?

(And `ToString()` in both callers, especially if you change the code paths that they would actually hit)

Perhaps. I might even agree with you, and I might have even proposed consistently using `CultureInfo.InvariantCulture` (or a caller-supplied `IFormatProvider`, as is consistent for framework APIs) if I had been reviewing the initial implementation of this before the initial release (knowing what I know today).

But why bury a locale-dependent behavior change in a PR of this nature? This is a public API that has stood for decades, and I personally believe that any behavior changes in such APIs should at least be given attention independently of many other **actually** minor changes.

File: test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs (hunk 1 above)
Comment (hahschaa) [reply test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs]: Okay, I just was wondering which side effects you mean, because I could not find any. I was just asking if I missed anything.

File: src/NetTopologySuite/Utilities/HexConverter.cs (hunk 2 above)
Comment (hahschaa) [reply src/NetTopologySuite/Utilities/HexConverter.cs:79]: Okay, thanks for the clarification. You are indeed correct that this is not the appropriate PR for it.
From my experience at work I know that things like this can lead to very weird bugs when communicating between systems with different locales. So that's why I changed it.

But like I said, you are correct and I will revert it.

File: test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs (hunk 1 above)
Comment (hahschaa) [reply test/NetTopologySuite.Tests.NUnit.Performance/Performance/Operation/Buffer/DepthSegmentStressTest.cs]: And it is not that I went out of my way to change this mindlessly. As far as I could see there are no side effects. That's why I made the change.

File: src/NetTopologySuite/Utilities/HexConverter.cs (hunk 2 above)
Comment (hahschaa) [reply src/NetTopologySuite/Utilities/HexConverter.cs:79]: Furthermore, yes I touches a lot of files, I'm aware. I just wanted to help do some maintenance.
I think even more could be done: There are indentation issues, style issues (e.g. some files prefix fields with an underscore, some fields don't, some fields use `m_`), etc. I can't help you much with the spatial algorithms as that is not my strong suit, but since I checked the code out anyway for trying to come up with a decent solution for [issues 790](https://github.com/NetTopologySuite/NetTopologySuite/issues/790) I thought I could help you with the simple stuff first.

**Comment by hahschaa** [comment]
I think the fence argument is a bit weird and over the top; with the same logic you could say that all the unused using statements were there for a reason and I am to dumb to understand why they are there. But the reality is just that someone forgot to clean up after removing code.

Besides, I did not make these changes willy-nilly, I actually checked if there were any side effects.

But I reverted the changes anyway.

**Review by airbreather — APPROVED** [review]

**Comment by airbreather** [comment]
The `license/cla` check is still pending.

@FObermaier as of right now, all of the changes are mechanical and/or require little-to-no actual thought. Should we insist on the CLA, or do you think it's OK to just merge? I personally think it's OK to just merge.

**IMO, don't complete with "Squash and merge"**. Each commit fixes a different thing, so it's IMO more useful for source control history to keep them separated. "Create a merge commit" is my preference here.

**Comment by hahschaa** [comment]
I have agreed to the CLA multiple times now. I have never done this before, so I might be doing something wrong here. I just clicked the button and then get

> You have agreed to the CLA for NetTopologySuite/NetTopologySuite
> 
> Version: 2018-10-10

Is is invalid after I push again?

**Comment by airbreather** [comment]
> I have agreed to the CLA multiple times now. I have never done this before, so I might be doing something wrong here. I just clicked the button and then get
> 
> > You have agreed to the CLA for NetTopologySuite/NetTopologySuite
> > Version: 2018-10-10
> 
> Is is invalid after I push again?

I see — in [the original comment from CLAassistant](https://github.com/NetTopologySuite/NetTopologySuite/pull/793#issuecomment-3069677013), it says this:

> **Harm van Schaaijk** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

That explains why the bot isn't detecting it. I do see your name on the list, so, merging.

Thanks for your contribution
